lib/ruby_lsp/index.rb

```diff
@@ -0,0 +1,63 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  class Index
+    extend T::Sig
+    include Singleton
+
+    sig { returns(T::Array[String]) }
+    attr_reader :files
+
+    sig { returns(Requests::Support::PrefixTree) }
+    attr_reader :prefix_tree
+
+    sig { void }
+    def initialize
+      @files = T.let(collect_files, T::Array[String])
+      @prefix_tree = T.let(Requests::Support::PrefixTree.new(@files), Requests::Support::PrefixTree)
+    end
+
+    sig { params(changes: T::Array[{ uri: String, type: Integer }]).void }
+    def synchronize(changes)
+      has_addition_or_removals = T.let(false, T::Boolean)
+
+      changes.each do |change|
+        # File change events include folders, but we're only interested in files
+        file_path = change[:uri].delete_prefix("file://")
```

Comment:
This will not always work properly. We should not be treating URIs as if they were plain strings, or we will get bitten by URL encoding problems.

For example `file:///some%20path/that%20includes%20spaces.txt` would fail this operation.

Reply:
I don't think you should need `Shellwords` at all, since we are not running anything in the shell. Ruby's `File` and `Pathname` classes can handle paths with spaces, etc inside them perfectly fine:
```ruby
[1] pry(main)> require "pathname"
=> false
[2] pry(main)> p = Pathname.new("some path/that includes spaces and an @ sign.txt")
=> #<Pathname:some path/that includes spaces and an @ sign.txt>
[3] pry(main)> p.parent.mkpath
=> nil
[4] pry(main)> p.write("foo bar baz")
=> 11
[5] pry(main)> p.read
=> "foo bar baz"
[6] pry(main)> .cat "some path/that includes spaces and an @ sign.txt"
foo bar baz
```